### ISSUE TITLE
Watch blacklist file to reload them automatically

### ISSUE THREAD

**Issue by Belphemur** [DNSCrypt/dnscrypt-proxy#714]
Hello,

Would it be possible to implement a watcher for the blacklist file ?

You have the [fsnotify](https://github.com/fsnotify/fsnotify) lib that seems to be quite easy to add to the project.

This way, you could watch for file change for the blacklist and simply reload them.

Because even now, with the new code, it's still slow when you have more than 500k entries. 

My use case is, I'm replacing pihole-FTL by dnscrypt-proxy since it's way more optimised in the term of memory usage for the same dataset of blacklist.

I'm also providing the service publicly and I don't really want to have any dns outage the time to restart the service for blacklist update.

ps: fsnotify also works on Windows.

**Comment by jedisct1** [comment]
Watching for file changes is trivial, but gracefully reloading configuration files is always a tricky thing to do.

You may have noticed a `Reload()` function in the plugin interface. This is exactly what it's here for. But currently, `Reload()` functions are empty.

It may do something like spawn a new goroutine that builds a new instance of the plugin, and then atomically replaces the previous one with the new one. And there's probably a generic way to do it for any plugin instead of reimplementing the same thing for every single plugin loading a configuration file.

It's been on my TODO list since the beginning, but I haven't found the time to do it. Could you possibly help?

**Comment by Belphemur** [comment]
Good point, I see it's not going to be the easiest system to implement, especially the logic behind each plugin for a graceful reload.

Another way to do it, supporting SIGHUP using [TableFlip](https://github.com/cloudflare/tableflip) of CloudFlare. They made a [whole article](https://blog.cloudflare.com/graceful-upgrades-in-go/) about their lib.

This way, the whole process is reloaded, all the config. They based their lib on how NGINX is doing the reload.

Eidt: I admit I have no idea how it would work with systemd socket.
And obviously the address bindings would still need a restart.

**Comment by jedisct1** [comment]
Tableflip also seems to require TCP connections. But DNS mostly uses UDP; there are no connections.
Also, it is only designed for Linux, whereas dnscrypt-proxy has to support many more operating systems, including Windows.

**Comment by Belphemur** [comment]
TableFlip support any socket that is provided by the network package of golang. Hence, UDP also supported. The idea been TableFlip is the main thread taking care of the network.

About Windows, golang does support the CTRL+C (^BREAK) of Windows. This could be the way to do the reload since it would only send a signal that is bindable. cf: [Windows Signal Golang](https://golang.org/pkg/os/signal/#hdr-Windows)

However, there is also a discussion here about how to do reload in Windows:
https://github.com/nats-io/gnatsd/issues/521

They suggest making the program a service in Windows (which make a lot of sense) linking to how NGINX work in Windows.

**Comment by jedisct1** [comment]
Interesting. But how does UDP work?

The proxy gets a query from a client, hasn't responded yet, an we want to update the configuration. 

Since there is no connection, how can tableflip know that this specific query shouldn't be handled by another instance yet?

**Comment by Belphemur** [comment]
After digging into the code of TableFlip (I'm still learning go, not an easy lib to read).

I'm not sure it could work with Windows as it used Unix Syscall to duplicate File Descriptor.

For what I gather, when you use the Listen on Fds in TableFlip, it will duplicate the Connection object and all the File Descriptor attached to it. It will also make it blocking. ([their tests](https://github.com/cloudflare/tableflip/blob/master/fds_test.go) for that component is giving quite a lot of use cases)

In other words, all UDP packet sent to the socket will be queued until the child shows up and start to read on it.

I'm thinking that going with the first idea of fsnotify on the plugin file would be a better idea, even if it means implementing a full reload logic that could be applied to any plugin. Not an easy task either.
